Improvement/advertise fetch optimize

sana7044 opened the pull request:
Fixed : Improvement/advertise fetch optimize

marclupanc commented on src/components/Advertiser/AdvertiserTable.vue:
```diff
@@ -273,7 +273,7 @@ const errorStore = useErrorStore()
 const selectedAdvertise = ref({})
 const filter = ref('')
-const selectedDataType = ref({ label: 'Active', value: 'active' })
+const selectedDataType = ref({ label: 'All', value: 'all' })
```
We already discussed this point on another ticket but you changed it back to fetch all advertises again. Is necessary at this point? 🤔

<img width="1070" alt="Screenshot 2024-08-29 at 17 11 24" src="https://github.com/user-attachments/assets/ee24fbac-e1be-4ed8-bd8f-fde377547e54">

sana7044 replied:
I though fetching all the data and filtering it locally would be much faster.

marclupanc replied:
Why than we won't just fetch all the documents for the whole app at once and then just use it locally? 😅 

It's not only about how would be faster, even though fetching 10 documents is much faster than fetching 100 documents. It's also about the cost of the reads/writes. [Please take a look](https://firebase.google.com/docs/firestore/pricing) 

Probaly currently we shouldn't be bothered by the amount of requests made by 2-3 devs in development branch, but it could grow in the future (while it is still 2 million reads for a month for few users, imagine having 1000 users). And if we can make it be more optimized, well organized and scalable now it will only help us in the future. 

So please make it to NOT fetch all the advertisments at once, and query only the necessary once. To notify user for a pending ad, we could also query that and show some badge in the future if that's the case.

![Screenshot 2024-09-05 at 11 55 14](https://github.com/user-attachments/assets/85133f55-32a5-4d31-ba9a-7cf177b8adc7)

sana7044 replied:
I am working on it

marclupanc commented on src/components/Advertiser/AdvertiserTable.vue:
```diff
@@ -644,10 +650,17 @@ async function onUpdate(e) {
   }
 }
-
-watch(selectedDataType, (newType) => {
```
So this watch is responsible for fetching only the selected advertises instead of fetching them all and then just filter them out by status. This is needed because of amount of requests are made to firestore. So basically our main goal is to optimize the app that way so we make as few requests as possible (make them only where its necessary) 

You can see the amount of requests being made to fetch all advertises VS to fetch only 1. Thats still a lot of requests even 1 advertisement, but it's less than to fetch all.

Imaginge we will have even more than that for all ads. 

https://github.com/user-attachments/assets/91743ceb-3300-4ccd-ab63-023fbdd904d9

https://github.com/user-attachments/assets/201804dd-04c0-4409-b650-6ff38578dab7

sana7044 replied:
These requests are being made in order to compute the total number of visits, impressions, and clicks for each advertisement.
Rather than recalculating each time we fetch an advertisement, we can update the cloud function to directly update the advertisement document with the total impressions, total clicks, and total visits. So that we don't need to fetch the impressions, clicks, or visits collection to compute total data for each advertisement.

marclupanc commented on src/stores/advertises.js:
```diff
@@ -71,26 +71,25 @@ export const useAdvertiseStore = defineStore('advertises', {
-            this.setLoaderFalse()
-          })
+          for (const advertise of advertises) {
+            advertise.author = userStore.getUserById(advertise.author.id) || (await userStore.fetchUser(advertise.author.id))
```
And I think for the advertisement, fetching all users for each advertisement is unecessery. 
(To test, remove this for loop and see that nothing changes in the advertisement table, but there are fewer requests)

If we need the user id for ad for editing we could fetch in separately by ad id.

marclupanc reviewed (changes requested):
@stravid87 Would like you to also take a look and tell your oppinion on this.

marclupanc commented:
Could you please provide more details (a list of things) you have worked on this PR so I could create another ticket for this PR as the #521 mentiones an issue which this PR doesn't resolve (on scroll there are still many requests made to update impressions, currently 100 requests to scroll down and up). So #521 will remain open for fixing the actual impressions issue in a separate PR so we can merge this to develop. 

https://github.com/user-attachments/assets/79d8720c-22a3-4974-9f0e-f7e4aeddb9fd

sana7044 commented:
> Could you please provide more details (a list of things) you have worked on this PR so I could create another ticket for this PR as the #521 mentiones an issue which this PR doesn't resolve (on scroll there are still many requests made to update impressions, currently 100 requests to scroll down and up). So #521 will remain open for fixing the actual impressions issue in a separate PR so we can merge this to develop.
> 

>  Screen.Recording.2024-09-09.at.17.29.551.mov

In this pr

- I have updated the cloud function so that it updates the advertisement doc totalImpressions, totalClicks, and totalVisits whenever an impression, clicks, or visits are added. So that we don't need to fetch all impressions, clicks, and visits data for each advertisement, we only need to fetch the advertisement collection.
- I have updated the advertisement table duration column to expiry status. It displays how many days are left, and when the advertisement expires, it displays expired instead of a negative number
